File: 0820-find-eventual-safe-states/0820-find-eventual-safe-states.py
class Solution:
    def eventualSafeNodes(self, graph: List[List[int]]) -> List[int]:
        n = len(graph)
        safe = [False] * n
        vis = [False] * n
        def dfs(node):
            vis[node] = True;
            safe[node] = True
            for i in graph[node]:
                if vis[i] == False and dfs(i) == True:
                    return True
                elif safe[i] == True:
                    return True
            safe[node] = False
            return False
        ans = []
        for i in range(0, n):
            if vis[i] == False:
                dfs(i)
        for i in range(0, n):
            if safe[i] == False:
                ans.append(i)
        return ans

File: 0820-find-eventual-safe-states/0820-find-eventual-safe-states.cpp
class Solution {
public:
    vector<int> eventualSafeNodes(vector<vector<int>>& graph) {
        int n = graph.size();
        vector<vector<int>> adj(n);
        vector<int> safe, ans;
        queue<int> q;
        for(int i = 0; i<n; i++) {
            for(auto a : graph[i]) adj[a].push_back(i);
            safe.push_back(graph[i].size());
            if(!safe[i]) q.push(i);
        }
        while(!q.empty()) {
            int node = q.front();
            q.pop();
            for(int i : adj[node]) {
                safe[i]--;
                if(!safe[i]) q.push(i);
            }
        }
        for(int i = 0; i<n; i++) if(!safe[i]) ans.push_back(i);
        return ans;
    }
};